### PR TITLE
Adjust run simulation

### PR DESCRIPTION
**Pull request by ostatni5**

### CODE REVIEW

File: src/WrapperApp/components/Simulation/SimulationPanel.tsx [hunk 1]
```diff
@@ -159,38 +159,6 @@ export default function SimulationPanel({
 		setInputFiles(inputFiles);
 	};
 
-	/**
```
Comment (grzanka) [inline src/WrapperApp/components/Simulation/SimulationPanel.tsx:162]: Why is this code removed ?

Comment (Derstilon) [reply src/WrapperApp/components/Simulation/SimulationPanel.tsx:162]: old function for running simulations. Replaced with sendSimulationRequest

**Comment by grzanka** [comment]
Bump JSON language version

**Comment by grzanka** [comment]
@ostatni5 still there are conflicts

**Review by grzanka — CHANGES_REQUESTED** [review]
@ostatni5 can you fix failing selenium tests ?

File: src/ThreeEditor/Simulation/Base/SimZone.ts [hunk 2]
```diff
@@ -142,6 +142,7 @@ export abstract class SimulationZone
 		const { uuid, name, materialUuid, materialPropertiesOverrides: overrides } = json;
 		this.uuid = uuid;
 		this.name = name;
+		console.log('lodain zone material', materialUuid);
```
Comment (grzanka) [inline src/ThreeEditor/Simulation/Base/SimZone.ts:145]: is this needed ?

File: src/services/ShSimulatorService.tsx [hunk 3]
```diff
@@ -236,6 +348,31 @@ const ShSimulation = ({ children }: ShSimulationProps) => {
...
+
+			const simData: ValidateShape<
+				typeof mergedData,
+				Partial<FullSimulationData>
+			> = mergedData;
+
+			return simData as FullSimulationData;
+		},
+		[getJobInputs, getJobResults, getJobStatus]
+	);
```
Comment (Derstilon) [inline src/services/ShSimulatorService.tsx:374]:
```suggestion
	const getFullSimulationData = useCallback(
		async (jobStatus: JobStatusData, signal?: AbortSignal, cache = true) => {
			const inputs: JobInputs | undefined = await getJobInputs(jobStatus, signal, cache);
			const results: JobResults | undefined =
				jobStatus.jobState === StatusState.COMPLETED
					? await getJobResults(jobStatus, signal, cache)
					: undefined;
			if (!inputs || !results) return undefined;
			const { message, ...mergedData } = {
				...inputs,
				...results,
				...jobStatus
			};

			const simData: FullSimulationData = mergedData satisfies ValidateShape<
				typeof mergedData,
				FullSimulationData
			>;

			return simData;
		},
		[getJobInputs, getJobResults]
	);
```

File: src/util/Types.ts [hunk 4]
```diff
@@ -1,3 +1,9 @@
 export type Common<A, B> = {
 	[P in keyof A & keyof B]: A[P] | B[P];
 };
+
+export type ValidateShape<T, Shape> = T extends Shape
+	? Exclude<keyof T, keyof Shape> extends never
+		? T
+		: never
+	: never;
```
Comment (Derstilon) [inline src/util/Types.ts:9]:
```suggestion
import { Equal } from '@mui/styled-engine';

export type Common<A, B> = {
	[P in keyof A & keyof B]: A[P] | B[P];
};

type ValidateKeys<T, Shape> = Equal<keyof T, keyof Shape, T, never>;

export type ValidateShape<T, Shape> = T extends Shape ? ValidateKeys<T, Shape> : never;
```

**Review by Derstilon — CHANGES_REQUESTED** [review]

File: src/WrapperApp/components/Simulation/SimulationPanel.tsx [hunk 5]
```diff
@@ -204,7 +186,7 @@ export default function SimulationPanel({
 		batchOptions: BatchOptionsType
 	) => {
 		setShowRunSimulationsForm(false);
-		const simData = sourceType === 'project' ? editorJson : inputFiles;
```
[the remark this answers is not in the document]
Comment (ostatni5) [reply src/WrapperApp/components/Simulation/SimulationPanel.tsx:207]: I think it is out of the scope of this PR. Create an issue for refactoring if you want.

**Review by Derstilon — COMMENTED** [review]
Update Readme in examples directory to include info about new migrations directory

Here is my suggestion



```md
# Examples directory

This directory contains examples of simulation projects for the editor. To add a new example, create a new JSON file in this directory. The naming convention for the file is as follows:

- ex{number of the example}.json

Please note that the examples must be numbered consecutively.

In addition to the examples, this directory also contains migration scripts for updating the structure of JSON files used in the simulation projects. The migration scripts are written in JavaScript and use the Node.js runtime environment. They are located in the `/migrations` directory.

To run this script using Node.js, follow these steps:

1. Open a terminal or command prompt.
2. Navigate to the `/migrations` directory.
3. Run the following command:

\```bash
node migrateEx<old_version>_<new_version>.js <filename>
\```

Replace `<old_version>` and `<new_version>` with the version numbers of the JSON files you want to migrate and the version you want to migrate to, respectively. 
Replace `<filename>` with the name of the JSON file you want to migrate.

For example, to migrate the `ex1.json` file from version `0.1` to version `0.2`, run the following command:

\```bash
node migrateEx0.1_0.2.js ex1.json
\```

The migrated file will be saved in the `/migrations` directory with the same name as the original file.
Backups of the original file will be saved in the `/migrations/temp` directory.

Note that not all version combinations are supported. 
The following table shows which versions can be migrated to which other versions:

| From version | To version |
| ------------ | ---------- |
| 0.7          | 0.9        |
```

File: src/ThreeEditor/examples/migrations/migrateEx0.7_0.9.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
+// Adjust this file to next migration version remember to update the version in the file name
+
+const path = require('path');
+const fs = require('fs');
+const fileNameToMigrate = process.argv[2];
+
+const pathToMigrate = path.resolve(process.cwd(), fileNameToMigrate);
+const fileJson = require(pathToMigrate);
+
```
Comment (Derstilon) [inline src/ThreeEditor/examples/migrations/migrateEx0.7_0.9.js:9]:
```suggestion

// create temp dir if not exists
if (!fs.existsSync('./migrations/temp'))
    fs.mkdirSync('./migrations/temp');

```

**Review by Derstilon — DISMISSED** [review]

**Review by Derstilon — APPROVED** [review]

**Review by grzanka — APPROVED** [review]

**Pull request merged**